File: Heap/Distant_Barcodes.py
#Problem Link:- https://leetcode.com/problems/distant-barcodes/

"""In a warehouse, there is a row of barcodes, where the ith barcode is barcodes[i].
Rearrange the barcodes so that no two adjacent barcodes are equal. You may return any 
answer, and it is guaranteed an answer exists."""

class Solution:
    def rearrangeBarcodes(self, barcodes: List[int]) -> List[int]:
        import heapq
        n = len(barcodes)
        d = dict()
        for i in range(n):
            if barcodes[i] not in d:
                d[barcodes[i]] = 1
            else:
                d[barcodes[i]] += 1
        dist = 0
        num = None
        heap = []
        ans = [None] * n
        for x in d:
            heapq.heappush(heap,(-d[x],x))
        for i in range(0,n,2):
            if dist == 0:
                dist,num = heapq.heappop(heap)
                dist = -dist
            ans[i] = num
            dist -= 1
        for i in range(1,n,2):
            if dist == 0:
                dist,num = heapq.heappop(heap)
                dist = -dist
            ans[i] = num
            dist -= 1
        return ans

File: Heap/DistantBarcodes.cpp
//Problem Link:- https://leetcode.com/problems/distant-barcodes/

#include<bits/stdc++.h>
using namespace std;

class Solution {
public:
    vector<int> rearrangeBarcodes(vector<int>& barcodes) {
        int n = barcodes.size();
        priority_queue<pair<int,int>> pq;
        unordered_map<int,int> ump;
        for(int i=0;i<n;i++){
            ump[barcodes[i]]++;
        }
        for(auto x : ump){
            pq.push({x.second,x.first});
        }
        int cnt = 0;
        int num;
        for(int i=0;i<n;i+=2){
            if (cnt == 0){
                cnt = pq.top().first;
                num = pq.top().second;
                pq.pop();
            }
            barcodes[i] = num;
            cnt--;
        }
        for(int i=1;i<n;i+=2){
            if (cnt == 0){
                cnt = pq.top().first;
                num = pq.top().second;
                pq.pop();
            }
            barcodes[i] = num;
            cnt--;
        }
        return barcodes;
    }
};